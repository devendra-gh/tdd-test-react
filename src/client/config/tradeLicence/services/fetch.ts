import baseFetch from 'universal-fetch';
import cookie from 'react-cookies';
import baseUrl from 'client/utils/baseUrl';

const defaults = {
  method: 'POST', // handy with GraphQL backends
  mode: 'same-origin',
  credentials: 'same-origin',
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
};

/**
 * @param {string} url
 * @param {string} method
 * @param {Object} data
 * @param {boolean} fileUpload
 * @returns {Promise<any | never>}
 */
export default async function api(
  url: string,
  method: string = 'GET',
  data?: object,
  fileUpload: boolean = false,
) {
  const config = {
    ...defaults,
    method,
    body: {},
  };

  if (method === 'POST' || method === 'PUT') {
    config.headers = {
      ...config.headers,
      // @ts-ignore
      'csrf-token': cookie.load('XSRF-TOKEN'),
    };
  }

  if (data && method !== 'GET') {
    if (!fileUpload) {
      config.headers = {
        ...config.headers,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      };
    }
    if (fileUpload) {
      delete config.headers['Content-Type'];
    }

    config.body = fileUpload ? data : JSON.stringify(data);
  } else {
    delete config.body;
  }

  try {
    const response = await baseFetch(`${baseUrl}${url}`, config);
    const payload = response.json();
    return payload;
  } catch (e) {
    throw e;
  }
}
