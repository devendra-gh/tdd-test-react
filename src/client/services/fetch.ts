import { IVariables } from '@tamm/app-composer';
import baseFetch from 'universal-fetch';
import cookie from 'react-cookies';
import baseUrl from '../utils/baseUrl';

interface IFetchConfig {
  method: string;
  mode: string;
  credentials: string;
  headers: IVariables;
}

const defaults: IFetchConfig = {
  method: 'POST', // handy with GraphQL backends
  mode: 'same-origin',
  credentials: 'same-origin',
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    'X-Requested-With': 'XMLHttpRequest',
  },
};
/**
 * @param {string} url
 * @param {string} method
 * @param {Object} data
 * @param {boolean} fileUpload
 * @returns {Promise<any | never>}
 */
export default async function api(
  url: string,
  method: string = 'GET',
  data?: object,
  fileUpload: boolean = false,
) {
  // eslint-disable-next-line no-param-reassign
  method = method.toUpperCase();
  const config = {
    ...defaults,
    method,
    body: {},
  };
  if (method === 'POST' || method === 'PUT') {
    config.headers = {
      ...config.headers,
      'csrf-token': cookie.load('XSRF-TOKEN'),
    };
  }
  if (data && method !== 'GET') {
    if (!fileUpload) {
      config.headers = {
        ...config.headers,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      };
    }
    if (fileUpload) {
      delete config.headers['Content-Type'];
    }
    config.body = fileUpload ? data : JSON.stringify(data);
  } else {
    delete config.body;
  }
  try {
    let response;
    if (url.includes('https://')) response = await baseFetch(url, data);
    else response = await baseFetch(`${baseUrl}${url}`, config);
    const payload = response.json();
    return payload;
  } catch (e) {
    throw e;
  }
}
