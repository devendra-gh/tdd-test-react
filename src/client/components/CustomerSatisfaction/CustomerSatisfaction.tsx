import React, { useCallback, useState } from 'react';

import CustomerSatisfactionComponent from '@tamm/ui-lib-v2-customer-satisfaction';

import fetch from 'client/services/fetch';

interface FeedbackProps {
  i18n: any;
  serviceId: string;
  productId?: string;
}

export function CustomerSatisfaction(props: FeedbackProps) {
  const [status, setStatus] = useState<
    'idle' | 'active' | 'pending' | 'success' | 'error'
  >('idle');
  const handleSubmit = useCallback(
    async (smileyType: string, options: Record<string, string>) => {
      setStatus('pending');
      try {
        await fetch('/pub/adfeedback/feedbacks', 'POST', {
          surveyType: 'service',
          smileyType,
          comments: options.comment,
          ratings: {
            clearness: options['customer_satisfaction.clear'],
            ease: options['customer_satisfaction.easy'],
            speed: options['customer_satisfaction.fast'],
            performance: options['customer_satisfaction.perf'],
          },
          pageUrl: window.location.href,
          serviceId: props.serviceId,
          productId: props.productId,
        });
        setStatus('success');
      } catch (e) {
        setStatus('error');
      }
    },
    [],
  );

  return (
    <CustomerSatisfactionComponent
      i18n={props.i18n}
      status={status}
      onSubmit={handleSubmit}
    />
  );
}
