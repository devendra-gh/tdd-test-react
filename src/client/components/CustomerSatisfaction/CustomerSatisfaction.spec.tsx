import React from 'react';
import { shallow } from 'enzyme';
import CustomerSatisfactionComponent from '@tamm/ui-lib-v2-customer-satisfaction';
import fetch from 'client/services/fetch';
import { CustomerSatisfaction } from './CustomerSatisfaction';

jest.mock('client/services/fetch');

describe('CustomerSatisfaction', () => {
  it('should render component', () => {
    const component = shallow(
      <CustomerSatisfaction
        i18n={() => ({})}
        serviceId="123"
        productId="123"
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('should render success state', async () => {
    const component = shallow(
      <CustomerSatisfaction
        i18n={() => ({})}
        serviceId="123"
        productId="123"
      />,
    );
    (fetch as jest.Mock).mockResolvedValue({});
    const onSubmit =
      component.find(CustomerSatisfactionComponent).prop('onSubmit') ||
      (() => {});
    await onSubmit('test', {});
    expect(component).toMatchSnapshot();
  });

  it('should render error state', async () => {
    const component = shallow(
      <CustomerSatisfaction
        i18n={() => ({})}
        serviceId="123"
        productId="123"
      />,
    );
    (fetch as jest.Mock).mockRejectedValue({});
    const onSubmit =
      component.find(CustomerSatisfactionComponent).prop('onSubmit') ||
      (() => {});
    await onSubmit('test', {});
    expect(component).toMatchSnapshot();
  });
});
