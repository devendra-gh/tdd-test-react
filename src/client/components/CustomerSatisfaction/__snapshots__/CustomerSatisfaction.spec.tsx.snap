// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerSatisfaction should render component 1`] = `
<CustomerSatisfaction
  emotion=""
  i18n={[Function]}
  onStatusChange={[Function]}
  onSubmit={[Function]}
  rates={
    Array [
      Object {
        "label": "customer_satisfaction.easy",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.clear",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.fast",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.perf",
        "value": 0,
      },
    ]
  }
  status="idle"
  withComment={true}
/>
`;

exports[`CustomerSatisfaction should render error state 1`] = `
<CustomerSatisfaction
  emotion=""
  i18n={[Function]}
  onStatusChange={[Function]}
  onSubmit={[Function]}
  rates={
    Array [
      Object {
        "label": "customer_satisfaction.easy",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.clear",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.fast",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.perf",
        "value": 0,
      },
    ]
  }
  status="error"
  withComment={true}
/>
`;

exports[`CustomerSatisfaction should render success state 1`] = `
<CustomerSatisfaction
  emotion=""
  i18n={[Function]}
  onStatusChange={[Function]}
  onSubmit={[Function]}
  rates={
    Array [
      Object {
        "label": "customer_satisfaction.easy",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.clear",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.fast",
        "value": 0,
      },
      Object {
        "label": "customer_satisfaction.perf",
        "value": 0,
      },
    ]
  }
  status="success"
  withComment={true}
/>
`;
